_posts/2019-09-19-dask-hyperparam-opt.md

```diff
@@ -0,0 +1,440 @@
+visualization tool, [t-SNE]. This tool requires (at least) three
+hyperparameters, and the first section in "[How to Use t-SNE
+Effectively][tsne-study]" is titled "Those hyperparameters really matter".
+
+[^regularization]:Performance comparison: Scikit-learn's visualization of tuning a Support Vector Machine's (SVM) regularization parameter: [Scaling the regularization parameter for SVCs][sklearn-reg]
+[tsne-study]:https://distill.pub/2016/misread-tsne/
+[sklearn-reg]:https://scikit-learn.org/stable/auto_examples/svm/plot_svm_scale_c.html
+[t-SNE]:https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html
+
+These hyperparameters need to be specified by the user. There are no good
+heuristics for determining what the values should be.  Finding decent values
+for these hyperparameters or "hyperparameter optimization" grows more complex
+as the number of hyperparameters grow.
+
+Finding good values for these hyperparameters is critical and has an entire
+Scikit-Learn documentation page, "[Tuning the hyperparameters of an
+estimator]." Briefly, finding the decent values of hyperparameters
+is difficult and requires guessing or searching.
+
+[Tuning the hyperparameters of an estimator]:http://scikit-learn.org/stable/modules/grid_search.html
+
+**How can these hyperparameters be found quickly and efficiently with an
+advanced task scheduler like Dask?** Parallelism will pose some challenges, but
+the Dask architecture enables some advanced algorithms.
+
+## Contributions
+
+Dask-ML can quickly find high-performing hyperparameters.
+
+I will back this
+claim with intuition and experimental evidence. Specifically, Dask-ML now
+implements an algorithm introduced by Li et. al. in "[Hyperband: A novel
+bandit-based approach to hyperparameter optimization][hyperband-paper]".
+
+[hyperband-paper]:https://arxiv.org/pdf/1603.06560.pdf
+
+Pairing of Dask and Hyperband enables some exciting new performance opportunities.
+Hyperband is fairly new,[^new] and this is the first implementation with an
+advanced task scheduler like Dask that I'm aware of.[^other-work]
+
+[^new]:By some metrics. It's been around since 2016 and some call that "old news."
+[^other-work]:Including with [Ray-Tune's implementation][asha-ray] and "[Massively Parallel Hyperparameter Tuning][asha]".
+[asha-ray]:https://ray.readthedocs.io/en/latest/tune-schedulers.html#asynchronous-hyperband
+[asha]:https://arxiv.org/pdf/1810.05934.pdf
+
+I'll highlight some key points of the corresponding paper,
+"[Better and faster hyperparameter optimization with Dask][scipy19]".  Let's go
+through the basics of Hyperband then illustrate it's use and performance with
+an example.
+
+## Hyperband basics
+
+The motivation for Hyperband is to high performing models with minimal
+training. Given this goal, it makes sense to spend more time training high
+performing models – why waste training time on poor performing models?
```

Comment:
May be worth introducing the incremental concept here? If we *knew* that it's a poor model, then we wouldn't need hyperparameter opt in the first place. But the intuition is that a model that starts out poor will probably be poor on future batches of data.

Reply:
Rewritten as

> training. Given this goal, it makes sense to spend more time training high performing models – why waste more time training time a model if it's done poorly in up until now?